File: app/src/main/java/com/peterlaurence/trekadvisor/core/map/maparchiver/MapArchiver.java
package com.peterlaurence.trekadvisor.core.map.maparchiver;

import com.peterlaurence.trekadvisor.core.map.MapArchive;
import com.peterlaurence.trekadvisor.util.UnzipTask;

import java.io.File;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

/**
 * Helper class to handler unzipping of a {@link MapArchive}.
 *
 * @author peterLaurence on 14/10/17.
 */
public class MapArchiver {
    /**
     * For instance, just unzips in a subfolder of the same parent folder of the archive
     * {@link File} passed as parameter. The subfolder is named from a formatting of the current
     * date.
     */
    public static void archiveMap(final MapArchive mapArchive, UnzipTask.UnzipProgressionListener listener) {
        /* Generate an output directory with the date */
        Date date = new Date();
        DateFormat dateFormat = new SimpleDateFormat("dd\\MM\\yyyy-HH:mm:ss", Locale.ENGLISH);
        String parentFolderName = mapArchive.getName() + "-" + dateFormat.format(date);
        File zipFile = mapArchive.getArchiveFile();
        File outputDirectory = new File(zipFile.getParentFile(), parentFolderName);

        /* Launch the unzip thread */
        UnzipTask unzipTask = new UnzipTask(zipFile, outputDirectory, listener);
        unzipTask.start();
    }
}

File: app/src/main/java/com/peterlaurence/trekme/core/map/maparchiver/MapArchiver.kt
package com.peterlaurence.trekme.core.map.maparchiver

import com.peterlaurence.trekme.util.UnzipProgressionListener
import com.peterlaurence.trekme.util.unzipTask
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.io.File
import java.io.InputStream
import java.text.SimpleDateFormat
import java.util.*

/**
 * Unzips the given [inputStream] (by creating a [ZipInputStream] and reading from it).
 * The output folder must be provided, along with:
 * * a name, which is used as prefix for the direct parent folder of the map. The current date is
 * appended.
 * * the size in bytes of the document being read
 *
 * @author peterLaurence on 28/02/20
 */
fun CoroutineScope.unarchive(inputStream: InputStream, outputDirectory: File, name: String, size: Long, listener: UnzipProgressionListener) {
    /* Generate an output directory with the date */
    val date = Date()
    val dateFormat = SimpleDateFormat("dd\\MM\\yyyy-HH:mm:ss", Locale.ENGLISH)
    val parentFolderName = name + "-" + dateFormat.format(date)
    val intermediateDirectory = File(outputDirectory, parentFolderName)

    /* Launch the unzip task */
    launch(Dispatchers.IO) {
        unzipTask(inputStream, intermediateDirectory, size, listener)
    }
}
